c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    const double *weight_row;
+    double n;
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        weight_row = GET_2D_ROW(hap_weights, 3, k);
+        n = (double) args.sample_set_sizes[k];
+        // TODO: what to do when n = 0
+        result[k] = weight_row[0] / n;
+    }
+    return 0;
+}
+
+static int
+norm_total_weighted(tsk_size_t state_dim, const double *TSK_UNUSED(hap_weights),
+    tsk_size_t n_a, tsk_size_t n_b, double *result, void *TSK_UNUSED(params))
+{
+    for (tsk_size_t k = 0; k < state_dim; k++) {
+        result[k] = 1 / (double) (n_a * n_b);
+    }
+    return 0;
+}
+
+static void
+get_all_samples_bits(tsk_bit_array_t *all_samples, tsk_size_t n, tsk_size_t n_chunks)
+{
+    const tsk_bit_array_t all = ~((tsk_bit_array_t) 0);
+    const tsk_bit_array_t remainder_samples = n % TSK_BIT_ARRAY_NUM_BITS;
+    all_samples[n_chunks - 1] = remainder_samples ? ~(all << remainder_samples) : all;
+    for (tsk_size_t i = 0; i < n_chunks - 1; i++) {
+        all_samples[i] = all;
+    }
+}
+
+static void
+sample_weights_to_bit_array(const double *weights, tsk_size_t num_sample_sets,
+    tsk_size_t num_samples, tsk_size_t num_sample_chunks, tsk_bit_array_t **sample_bits)
+{
+    const double *weight_row;
+    tsk_bit_array_t *bits_row;
+    for (tsk_bit_array_t i = 0; i < num_samples; i++) {
+        weight_row = GET_2D_ROW(weights, num_sample_sets, i);
+        for (tsk_size_t k = 0; k < num_sample_sets; k++) {
             if (weight_row[k] != 0) {
-                ret = TSK_ERR_DUPLICATE_SAMPLE;
+                bits_row = GET_2D_ROW(*sample_bits, num_sample_chunks, k);
+                tsk_add_bit_to_bit_array(bits_row, i);
+            }
+        }
+    }
+}
+
+typedef int norm_func_t(tsk_size_t state_dim, const double *hap_weights, tsk_size_t n_a,
+    tsk_size_t n_b, double *result, void *params);
+
+// TODO: should site_a/b be an id_t?
+static int
+compute_general_two_site_stat_result(tsk_size_t site_a, tsk_size_t site_a_offset,
+    tsk_size_t site_b, tsk_size_t site_b_offset, tsk_size_t num_sample_chunks,
+    const tsk_size_t *num_alleles, const tsk_bit_array_t *state, tsk_size_t state_dim,
+    tsk_bit_array_t *sample_sets, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f, bool polarised,
+    double *result)
+{
+    int ret = 0;
+    const tsk_bit_array_t *A_samples, *B_samples;
+    tsk_size_t w_A = 0, w_B = 0, w_AB = 0;
+    tsk_bit_array_t *ss_row; // ss_ prefix refers to a sample set
+    tsk_bit_array_t ss_A_samples[num_sample_chunks], ss_B_samples[num_sample_chunks],
+        ss_AB_samples[num_sample_chunks], AB_samples[num_sample_chunks];
+    uint8_t polarised_val = polarised ? 1 : 0;
+
+    double weights[3 * state_dim];
+    double norm[state_dim];
+    double *hap_weight_row;
+    tsk_size_t row_len = num_alleles[site_b] * state_dim;
+    // TODO: is this stack allocation dangerous??
```

Comment:
Beyond using stack memory like this, it seems as though windows doesn't like my use of VLAs. I'm not sure if this is an inherent issue with MSVC or an issue with the c standard definition in the build. I see `ignoring unknown option '-std=c99'` in the build logs here: https://github.com/tskit-dev/tskit/actions/runs/5592422530/job/15156184424?pr=2805#step:12:35
It does seem as though MVSC is compliant up until C17, so I'd expect there to be support for VLAs, but maybe I'm not understanding the issue.

Reply:
AFAIK, MVSC supports thru at least C11 but **without** VLAS.  Some Googling seems to agree, at least as of 2021: https://www.reddit.com/r/C_Programming/comments/kzskyi/why_cant_use_a_variable_to_declare_an_arrays_size/

Reply:
Another thing to consider: VLAs are legal, but rarely used in C.  They are **not** valid C++ code, and tskit may be compiled with a C++ compiler. Doing so is probably rare in practice, but we don't want to prevent it.

Reply:
Ah, good to know. Thanks @molpopgen I think I'll have to rework this a bit to remove the reliance on VLAs. Since it's in a pretty hot loop, I was thinking about pre-allocating some memory and passing it in.

Reply:
I haven't looked in detail at your changes, but we can certainly start simple and run `perf` once tests are passing, etc., to see where improvements are likely.

Reply:
My 10c is that allocing anything on the  stack that isn't a fixed, small size just isn't worth the hassle. It's not portable, unpredictable when it fails and doesn't give much/any perf advantages.